### PR TITLE
MG-2287 - Improve Search For Things

### LINKED ISSUE #2287: Improve search for Users, Things, Channels, and  Groups
We need to enable backend support for filtering _by name_. In the initial implementation here https://github.com/absmach/magistrala/blob/3355332ebdcc8bfd0b35ef8915a3734462f40a3f/pkg/clients/postgres/clients.go#L194 we return _id, name, creation and update time_, but we need only _id and name_. Let's also rename to `SearchBasicInfo` or something similar to avoid association with List (`RetrieveAll`) operations. We also need it across different services for Thing and Channels (when assigning to Channel or filtering Things on the UI, for example).

### PR DESCRIPTION
**Pull request by nyagamunene**
# What type of PR is this?
This is a feature because it improve search for Things

## What does this do?
This enables back-end support for filtering by name, ID and tags.

## Which issue(s) does this PR fix/relate to?

- Related Issue #2287 
- Resolves #2287 

## Have you included tests for your changes?
Yes, I have included tests for my changes.

## Did you document any new/modified feature?
No.  I have not updated the documentation because it will be done in a different PR

### CODE REVIEW

File: pkg/clients/postgres/clients.go [hunk 1]
```diff
@@ -190,7 +190,7 @@ func (repo *Repository) RetrieveAll(ctx context.Context, pm clients.Page) (clien
 	return page, nil
 }
 
-func (repo *Repository) RetrieveAllBasicInfo(ctx context.Context, pm clients.Page) (clients.ClientsPage, error) {
+func (repo *Repository) SearchBasicInfo(ctx context.Context, pm clients.Page) (clients.ClientsPage, error) {
```
Comment (arvindh123) [inline pkg/clients/postgres/clients.go]: I think this change is not required, since we have `SearchBasicInfo` function in things postgres

File: pkg/clients/clients.go [hunk 2]
```diff
@@ -78,8 +78,8 @@ type Repository interface {
 	// RetrieveAll retrieves all clients.
 	RetrieveAll(ctx context.Context, pm Page) (ClientsPage, error)
 
-	// RetrieveAllBasicInfo list all clients only with basic information.
-	RetrieveAllBasicInfo(ctx context.Context, pm Page) (ClientsPage, error)
+	// SearchBasicInfo list all clients only with basic information.
+	SearchBasicInfo(ctx context.Context, pm Page) (ClientsPage, error)
```
Comment (arvindh123) [inline pkg/clients/clients.go]: I think this change is not required, since we have SearchBasicInfo function in things postgres

**Review by arvindh123 — CHANGES_REQUESTED** [review]

File: pkg/clients/postgres/clients.go (hunk 1 above)
Comment (nyagamunene) [reply pkg/clients/postgres/clients.go]: From the comment left by @dborovcanin [here](https://github.com/absmach/magistrala/issues/2287#issue-2349057815), I understood that the renaming is necessary to avoid confusion between retrieve and search.

Comment (arvindh123) [reply pkg/clients/postgres/clients.go]: Yeah, I agree 
But the RetrieveAllBasicInfo will be no more needed , it will be replaced by search.

Comment (nyagamunene) [reply pkg/clients/postgres/clients.go]: Okay then

File: things/postgres/clients.go [hunk 3]
```diff
@@ -6,7 +6,9 @@ package postgres
 import (
 	"context"
 	"fmt"
+	"strings"
 
+	"github.com/absmach/magistrala/internal/api"
```
Comment (dborovcanin) [inline things/postgres/clients.go]: Avoid importing API to repo even if it's internal.

File: things/postgres/clients.go [hunk 4]
```diff
@@ -33,6 +35,12 @@ type Repository interface {
 
 	// RetrieveBySecret retrieves a client based on the secret (key).
 	RetrieveBySecret(ctx context.Context, key string) (mgclients.Client, error)
+
+	// Delete deletes client with given id
+	Delete(ctx context.Context, id string) error
```
Comment (dborovcanin) [inline things/postgres/clients.go]: This change does not belong to this PR.

File: things/postgres/clients.go [hunk 5]
```diff
@@ -121,3 +129,96 @@ func (repo clientRepo) RetrieveBySecret(ctx context.Context, key string) (mgclie
...
+func constructSearchQuery(pm mgclients.Page) (string, string) {
+	var query []string
+	var emq string
+	var tq string
+
+	if pm.Name != "" {
+		query = append(query, "name ~* :name")
+	}
+	if pm.Identity != "" {
+		query = append(query, "id ~* :identity")
```
Comment (dborovcanin) [inline things/postgres/clients.go]: There is no need to use regular expresions, since we do not have an advanced query. LIKE or ILIKE should be a better fit and faster.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: things/service.go [hunk 6]
```diff
@@ -530,6 +530,30 @@ func (svc service) ListClientsByGroup(ctx context.Context, token, groupID string
 	}, nil
 }
 
+func (svc service) SearchThings(ctx context.Context, token string, pm mgclients.Page) (mgclients.ClientsPage, error) {
+	res, err := svc.identify(ctx, token)
+	if err != nil {
+		return mgclients.ClientsPage{}, err
+	}
+
+	cp, err := svc.clients.SearchBasicInfo(ctx, pm)
```
Comment (arvindh123) [inline things/service.go]: @nyagamunene 

Instead of `SearchBasicInfo` can rename to `SearchClients` ?

File: things/service.go [hunk 7]
```diff
@@ -181,7 +181,7 @@ func (svc service) ListClients(ctx context.Context, token, reqUserID string, pm
 
 	pm.IDs = ids
 
-	tp, err := svc.clients.RetrieveAllByIDs(ctx, pm)
+	tp, err := svc.SearchThings(ctx, token, pm)
```
Comment (arvindh123) [inline things/service.go]: Why here we are calling `SearchThing` function
And not calling repo function `svc.clients.SearchBasicInfo(ctx, pm)`
`

**Review by arvindh123 — CHANGES_REQUESTED** [review]

File: things/service.go (hunk 7 above)
Comment (nyagamunene) [reply things/service.go]: In the SearchThings were are checking if the user has authority to view the thing so it is separated from `listclients` when retrieving the basic info

File: things/service.go (hunk 6 above)
Comment (nyagamunene) [reply things/service.go]: We can rename it to `SearchClients` no problem

File: things/service.go (hunk 7 above)
Comment (arvindh123) [reply things/service.go]: Authorization are done on above switch case.

File: things/service.go [hunk 8]
```diff
@@ -530,6 +530,30 @@ func (svc service) ListClientsByGroup(ctx context.Context, token, groupID string
...
+		return mgclients.ClientsPage{}, err
+	}
+
+	things := mgclients.ClientsPage{}
+	for _, val := range cp.Clients {
+		if _, err := svc.authorize(ctx, res.GetDomainId(), auth.UserType, auth.UsersKind, res.GetId(), auth.ViewPermission, auth.ThingType, val.ID); err == nil {
+			things.Clients = append(things.Clients, val)
+			things.Page.Total += 1
+		}
+	}
```
Comment (arvindh123) [inline things/service.go]: This will not return proper page items.

Comment (arvindh123) [reply things/service.go]: I think we don't need this `SearchThings` function in service layer

**Review by arvindh123 — CHANGES_REQUESTED** [review]

File: things/service.go (hunk 8 above)
Comment (nyagamunene) [reply things/service.go]: I will remove the `SearchThings` function then.

Comment (nyagamunene) [reply things/service.go]: The authorization is needed because the `member` should not be able to `view` things created by other members

Comment (arvindh123) [reply things/service.go]: Yeah, In `ListThings`  it is handled.

File: pkg/clients/postgres/clients.go [hunk 9]
```diff
@@ -537,6 +519,12 @@ func ConstructSearchQuery(pm clients.Page) (string, string) {
 	if pm.Id != "" {
 		query = append(query, "id ILIKE '%' || :id || '%'")
 	}
+	if pm.Tag != "" {
+		query = append(query, "EXISTS (SELECT 1 FROM unnest(tags) AS tag WHERE tag ILIKE '%' || :tag || '%')")
+	}
+	if len(pm.IDs) != 0 {
+		query = append(query, fmt.Sprintf("id IN ('%s')", strings.Join(pm.IDs, "','")))
+	}
```
Comment (arvindh123) [inline pkg/clients/postgres/clients.go]: Something like this will make code reusable and make simpler.
```suggestion
func buildQueryConditions(pm clients.Page) ([]string) {
	var query []string

	if pm.Name != "" {
		query = append(query, "name ILIKE '%' || :name || '%'")
	}
	if pm.Identity != "" {
		query = append(query, "identity ILIKE '%' || :identity || '%'")
	}
	if pm.Id != "" {
		query = append(query, "id ILIKE '%' || :id || '%'")
	}
	if pm.Tag != "" {
		query = append(query, "EXISTS (SELECT 1 FROM unnest(tags) AS tag WHERE tag ILIKE '%' || :tag || '%')")
	}
	if len(pm.IDs) != 0 {
		query = append(query, fmt.Sprintf("id IN ('%s')", strings.Join(pm.IDs, "','")))
	}

	

	return query
}

func applyOrdering(emq string, pm clients.Page) string {
	switch pm.Order {
	case "name", "identity", "created_at", "updated_at":
		emq = fmt.Sprintf("%s ORDER BY %s", emq, pm.Order)
		if pm.Dir == api.AscDir || pm.Dir == api.DescDir {
			emq = fmt.Sprintf("%s %s", emq, pm.Dir)
		}
	}
	return emq
}

func PageQuery(pm clients.Page) (string, error) {
	mq, _, err := postgres.CreateMetadataQuery("", pm.Metadata)
	if err != nil {
		return "", errors.Wrap(errors.ErrMalformedEntity, err)
	}

	query := buildQueryConditions(pm)
	if mq != "" {
		query = append(query, mq)
	}

	// Additional conditions specific to PageQuery
	if pm.Status != clients.AllStatus {
		query = append(query, "c.status = :status")
	}
	if pm.Domain != "" {
		query = append(query, "c.domain_id = :domain_id")
	}
	if pm.Role != clients.AllRole {
		query = append(query, "c.role = :role")
	}

	if len(query) > 0 {
		emq = fmt.Sprintf("WHERE %s", strings.Join(query, " AND "))
	}

	var emq string
	if len(query) > 0 {
		emq = fmt.Sprintf("WHERE %s", strings.Join(query, " AND "))
	}
	emq = applyOrdering(emq, pm)
	return emq, nil
}

func ConstructSearchQuery(pm clients.Page) (string) {
	query := buildQueryConditions(pm)
	var emq string
	if len(query) > 0 {
		emq = fmt.Sprintf("WHERE %s", strings.Join(query, " AND "))
	}
	emq = applyOrdering(emq, pm)
	return emq, tq
}
```

**Review by arvindh123 — CHANGES_REQUESTED** [review]

File: things/api/http/endpoints_test.go [hunk 10]
```diff
@@ -364,280 +364,6 @@ func TestCreateThings(t *testing.T) {
 	}
 }
 
-func TestListThings(t *testing.T) {
```
Comment (dborovcanin) [inline things/api/http/endpoints_test.go]: Why is this test removed?

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: pkg/clients/postgres/clients.go [hunk 11]
```diff
@@ -487,47 +469,38 @@ func PageQuery(pm clients.Page) (string, error) {
...
+	var emq string
 	if len(query) > 0 {
 		emq = fmt.Sprintf("WHERE %s", strings.Join(query, " AND "))
 	}
 	return emq, nil
 }
 
-func ConstructSearchQuery(pm clients.Page) (string, string) {
-	var query []string
+func constructSearchQuery(pm clients.Page) (string, string) {
```
Comment (dborovcanin) [inline pkg/clients/postgres/clients.go]: This function is used only once. Remove it and put the code directly to the `SearchClients`.

File: pkg/clients/postgres/clients.go [hunk 12]
```diff
@@ -537,20 +510,26 @@ func ConstructSearchQuery(pm clients.Page) (string, string) {
 	if pm.Id != "" {
 		query = append(query, "id ILIKE '%' || :id || '%'")
 	}
-
-	if len(query) > 0 {
-		emq = fmt.Sprintf("WHERE %s", strings.Join(query, " AND "))
+	if pm.Tag != "" {
+		query = append(query, "EXISTS (SELECT 1 FROM unnest(tags) AS tag WHERE tag ILIKE '%' || :tag || '%')")
+	}
+	if len(query) == 0 && len(pm.IDs) != 0 {
```
Comment (dborovcanin) [inline pkg/clients/postgres/clients.go]: We are treating IDs differently, i.e. taking it into account only if other criteria is not there, making `buildQueryConditions` mask the logic of processing IDs. Why?

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: pkg/clients/postgres/clients.go (hunk 12 above)
Comment (nyagamunene) [reply pkg/clients/postgres/clients.go]: This is allows `searchClients` to retrieve basic info by id or search. So if either `name`, `tag`, `identify`, or `id` is set we don't retrieve by `id`

File: pkg/clients/postgres/clients.go (hunk 11 above)
Comment (nyagamunene) [reply pkg/clients/postgres/clients.go]: Okay

File: pkg/clients/postgres/clients.go (hunk 12 above)
Comment (dborovcanin) [reply pkg/clients/postgres/clients.go]: Why don't we still combine? If you want only IDs, you send only IDs in the query.

Comment (nyagamunene) [reply pkg/clients/postgres/clients.go]: We can but to get you properly, you are suggesting we combine both `searchClients` and `retrieveAllByIds` right?

Comment (dborovcanin) [reply pkg/clients/postgres/clients.go]: Eventually, that would be the outcome here, yes. I.e. Search users from this pool of IDs with this criteria. It's unlikely to need such a thing, but it makes `buildQueryConditions` consistent.

Comment (arvindh123) [reply pkg/clients/postgres/clients.go]: @dborovcanin  @nyagamunene 
For search with ID, it will work even the ID matches partially. 
So we need list of things to which user have access.
As you know, Things service could not have information of user accessiable thing, so it will retreive from spiceDB for all the IDs to which user have access. 

Once we find way to list things efficently with access control, then we can remove this IDs from entities

File: pkg/clients/postgres/clients.go [hunk 13]
```diff
@@ -487,47 +475,27 @@ func PageQuery(pm clients.Page) (string, error) {
...
 		query = append(query, "c.status = :status")
 	}
 	if pm.Domain != "" {
 		query = append(query, "c.domain_id = :domain_id")
 	}
-
-	if pm.Role != clients.AllRole {
-		query = append(query, "c.role = :role")
-	}
+	var emq string
```
Comment (arvindh123) [inline pkg/clients/postgres/clients.go]: At present `RetrieveAll` and `RetrieveByIds` doesn't have ordering . 
I think it is good to have ordering in all list and search function.

**Review by arvindh123 — CHANGES_REQUESTED** [review]

File: pkg/clients/postgres/clients.go (hunk 13 above)
Comment (nyagamunene) [reply pkg/clients/postgres/clients.go]: I can add that

File: users/service.go [hunk 14]
```diff
@@ -193,10 +194,11 @@ func (svc service) ListClients(ctx context.Context, token string, pm mgclients.P
 		Offset:   pm.Offset,
 		Limit:    pm.Limit,
 		Name:     pm.Name,
+		Id:       pm.Id,
 		Identity: pm.Identity,
 		Role:     mgclients.UserRole,
 	}
-	pg, err := svc.clients.SearchBasicInfo(ctx, p)
+	pg, err := svc.clients.SearchClients(ctx, p)
```
Comment (arvindh123) [inline users/service.go]: Why do we need to pulicate the mgclients.Page, instead we can pass variable pm directly to SearchClients
```suggestion
    pg, err := svc.clients.SearchClients(ctx, pm)
```

**Review by arvindh123 — CHANGES_REQUESTED** [review]

File: pkg/clients/postgres/clients.go [hunk 15]
```diff
@@ -487,47 +477,46 @@ func PageQuery(pm clients.Page) (string, error) {
...
+	case "name", "identity", "created_at", "updated_at":
+		emq = fmt.Sprintf("%s ORDER BY %s", emq, pm.Order)
+		if pm.Dir == api.AscDir || pm.Dir == api.DescDir {
+			emq = fmt.Sprintf("%s %s", emq, pm.Dir)
+		}
+	}
+	return emq
+}
 
+func buildSearch(query []string, pm clients.Page) string {
```
Comment (dborovcanin) [inline pkg/clients/postgres/clients.go]: Remove this function and streamline `PageQuery` to something like: 
```suggestion
func PageQuery(pm clients.Page) (string, error) {
	mq, _, err := postgres.CreateMetadataQuery("", pm.Metadata)
	if err != nil {
		return "", errors.Wrap(errors.ErrMalformedEntity, err)
	}

	var query []string

	if pm.Name != "" {
		query = append(query, "name ILIKE '%' || :name || '%'")
	}
	if pm.Identity != "" {
		query = append(query, "identity ILIKE '%' || :identity || '%'")
	}
	if pm.Id != "" {
		query = append(query, "id ILIKE '%' || :id || '%'")
	}
	if pm.Tag != "" {
		query = append(query, "EXISTS (SELECT 1 FROM unnest(tags) AS tag WHERE tag ILIKE '%' || :tag || '%')")
	}
	if pm.Role != clients.AllRole {
		query = append(query, "c.role = :role")
	}
	// If there are search params presents, use search and ignore other options.
	// Always combine role with search params, so len(query) > 1.
	if len(query) > 1 {
		return fmt.Sprintf("WHERE %s", strings.Join(query, " AND ")), nil
	}

	if mq != "" {
		query = append(query, mq)
	}
	if len(pm.IDs) != 0 {
		query = append(query, fmt.Sprintf("id IN ('%s')", strings.Join(pm.IDs, "','")))
	}
	if pm.Status != clients.AllStatus {
		query = append(query, "c.status = :status")
	}
	if pm.Domain != "" {
		query = append(query, "c.domain_id = :domain_id")
	}

	var emq string
	if len(query) > 0 {
		emq = fmt.Sprintf("WHERE %s", strings.Join(query, " AND "))
	}
	return emq, nil
}
```
Don't forget comments, it's important because the function is a little cryptic. We must improve, I have opened a ticket for this: #2326.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

**Review by dborovcanin — APPROVED** [review]
